OAuth 2.0 Grant Type SPI

dteleguin opened the pull request:
Closes: #26250

dteleguin committed:
OAuth 2.0 Grant Type SPI

Closes: #26250

Signed-off-by: Dmitry Telegin <demetrio@carretti.pro>

pedroigor commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java:
```diff
@@ -267,34 +189,12 @@ private void checkGrantType() {
+
+        event.event(grant.getEventType());
+
+        if (grantType.equals(OAuth2Constants.DEVICE_CODE_GRANT_TYPE)
```
Any reason to not rely on `EnvironmentDependentProviderFactory` and avoid this check?

dteleguin replied:
Good point. Updated the code to use `EnvironmentDependentProviderFactory` for CIBA, Device Code and Token Exchange.

The side effect of this will be Keycloak returning `400 Bad Request` instead of `501 Not Implemented` for disabled grant types, as we will no longer be calling `ProfileHelper.requireFeature()`.

pedroigor commented on services/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantTypeBase.java:
```diff
@@ -0,0 +1,278 @@
+    protected HttpResponse response;
+    protected HttpHeaders headers;
+
+    protected void initialize(Context context) {
```
Do we need to call this method from every single implementation? Doesn't make more sense to have each implementation defining their fields (even if they need the same set)?

It will make the methods in this base class more self-contained?

dteleguin replied:
In regards to the fields in OAuth2GrantTypeBase, they all need to be there because the class contains common methods that depend on that fields.

But I agree about `initialize()` calls being redundant. I have replaced it with `setContext()` that will be called only once from `TokenEndpoint`.

pedroigor commented on server-spi-private/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantType.java:
```diff
@@ -0,0 +1,122 @@
+     *
+     * @return grant type name
+     */
+    String getGrantType();
```
What if we also support this https://github.com/keycloak/keycloak/pull/24512 from this SPI?

pedroigor reviewed (commented):
@dteleguin Thanks for starting this :)

pedroigor commented:
@dteleguin Still related to the comment around not binding providers to a single grant type identifier but allow providers to actually check whether they support or not processing a grant type request.

As discussed in our last meeting, it should also help decompose a grant type into multiple and more concise implementations. For instance, token exchange could provide each supported semantic (internal-internal, external-internal, delegation, impersonation, etc) by having their own implementation. Rather than a single provider covering all of them. 

The same for RFC7523.

dteleguin committed:
Use EnvironmentDependentProviderFactory for DeviceGrantType

dteleguin committed:
- rework grant type resolution to use supports() in addition to grant type
- replace initialize() with setContext()
- use EnvironmentDependentProviderFactory instead of runtime checks
- move OAuth2GrantTypeManager to server-spi-private
- javadocs, imports, minor fixes

Signed-off-by: Dmitry Telegin <demetrio@carretti.pro>

dteleguin commented:
> @dteleguin Still related to the comment around not binding providers to a single grant type identifier but allow providers to actually check whether they support or not processing a grant type request.

> As discussed in our last meeting, it should also help decompose a grant type into multiple and more concise implementations. For instance, token exchange could provide each supported semantic (internal-internal, external-internal, delegation, impersonation, etc) by having their own implementation. Rather than a single provider covering all of them.
> 
> The same for RFC7523.

I have checked out #24512 (thanks @ben95cd!). Looks like it borrows the idea from Token Exchange (introducing its own SPI and using a context wrapper + `boolean supports()` method to select the best provider).

If I get it right, your idea is to get rid of grant-specific SPIs (TokenExchangeSpi, AssertionGrantSpi etc.) and move the context wrapper + `supports()` logic to the upper level (TokenEndpoint). Using that approach, concrete Token Exchange variants could simply implement `OAuth2GrantType` with grant type set to `urn:ietf:params:oauth:grant-type:token-exchange` but different logic in `supports()` allowing provider selection; the same for Assertion Grant variants.

With that in mind, I have reworked grant type resolution to use two-phased approach. I'm afraid we can't get rid of grant type matching completely; for performance reasons, we wouldn't want `supports()` to be called for every registered sub-provider, but only for those that match the given grant type. So the first phase (in `checkGrantType()`) selects candidate implementations using grant type matching, then the second phase (in `resolveGrantType()`) narrows that list down to a concrete implementation. Unfortunately this could not be done in a single pass because client resolution needs to happen in between.

mposolda reviewed (commented):
@dteleguin Thanks! Added few comments on top of Pedro's comments. Hope we can have this available soon (before some conflicts are introduced in the meantime in `TokenEndpoint` :-) )

dteleguin commented on server-spi-private/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantManager.java:
```diff
@@ -0,0 +1,110 @@
+     * @return list of implementations
+     */
+    public static List<OAuth2GrantType> resolve(String grantType) {
+        return GRANT_TYPE_MAP.get(grantType);
```
The point of having `GRANT_TYPE_MAP` is to ensure performance and robustness with the growing number of Grant Type implementations.

As far as I understand, our plan is to get rid of grant type specific SPIs (Token Exchange SPI, Assertion Grant SPI etc.; please correct me @pedroigor if I'm wrong) and to migrate to one unified Grant Type SPI instead. This will already result in 7 grant types + 3 Token Exchange variants, at the very least. Considering the number of upcoming specifications that rely on Token Exchange and new grant types, one day we can easily end up with 15+ built-in grants, plus customers adding 1-2 on top of that. In such case, cycling over `supports()` for _every_ token request will be an overkill. Moreover, one single poorly designed implementation could slow down or even break the whole token issuance process, DoSing the server. This is why I suggested to filter the candidates using grant type name (string), then refine the selection using `supports()`.

As for potential hot-reload, I don't think this should be problematic since factories will register themselves in the map during loading (`postInit()`) and de-register when unloading (`close()`), so the map is fully dynamic.

But I agree that the PR is already complex. For now, we can use a simplified approach that you've suggested, and then add two-phase resolution as a follow-up, WDYT?

mposolda replied:
I think using `id` for grant type would be nice. This means that we don't need `supports()` method, which could mean further simplification?

I am not sure if there are some corner cases (like for example if there is some specification supporting something like "dynamic grant" , which means that `grant_type` can be any parameter in the format of like `urn:foo:bar:(.*)` . In that case `id` won't work and we would need `supports()` ? But if existing grants don't need that, we can perhaps just rely on id?

dteleguin commented on server-spi-private/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantType.java:
```diff
@@ -0,0 +1,139 @@
+ *
+ * @author <a href="mailto:demetrio@carretti.pro">Dmitry Telegin</a>
+ */
+public interface OAuth2GrantType extends Provider, ProviderFactory<OAuth2GrantType> {
```
Agreed. I have modeled it after Authenticators, but now I see that those are singletons without any local state, which is not the case for Grant Types.

What if we split the interface into `OAuth2GrantTypeFactory` + `OAuth2GrantType`, but then move factories into public inner classes, in order not to double the class number? like e.g. `AuthorizationCodeGrantType.Factory`?

mposolda replied:
@dteleguin Thanks for addressing this point! I am resolving.

pedroigor commented on services/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantTypeBase.java:
```diff
@@ -0,0 +1,283 @@
+
+    @Override
+    public void postInit(KeycloakSessionFactory factory) {
+        OAuth2GrantManager.register(this);
```
Related to the commend from @mposolda here https://github.com/keycloak/keycloak/pull/26251/files#r1472472745.

This is kinda against how providers/factories should be installed/inited. Now in addition to the usual installation of providers/factories (when the session factory is created and in Quarkus during re-aug) you are introducing another form of installing providers.

We should just rely on the existing mechanisms.

Another problem I see with this is the possibility of arbitrary extensions suddenly registering providers at runtime.

mposolda commented on server-spi-private/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantType.java:
```diff
@@ -0,0 +1,139 @@
+     */
+    Response process();
+
+    public static class Context {
```
+1, I was wondering about the same.

Also I think that most of the fields on the `Context` class can be `final` as they are sent in the constructor (in case we need the fields)

Finally, it can be also good if those fields are `private` and available via getters.

However I also understand that you likely wanted to avoid lots of changes in existing code when accessing those variables and that is why you introduced this?

But at least `final` on the fields can be low-hanging fruit without requiring any major changes IMO?

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java:
```diff
@@ -195,31 +133,18 @@ public Response processGrantRequest() {
 
-        throw new RuntimeException("Unknown action " + action);
+        resolveGrantType();
+        grant.setContext(context);
```
+1

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java:
```diff
@@ -267,40 +192,19 @@ private void checkGrantType() {
             throw newUnsupportedGrantTypeException();
         }
 
+        event.event(OAuth2GrantManager.grantToEvent(grantType));
```
+1

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/grants/AuthorizationCodeGrantType.java:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    @Override
+    public String getId() {
```
+1 (as long as we don't need to support complex / dynamic grant types as I've mentioned in the other comment. But if we need, we can perhaps adjust later if needed? So I would personally consider just if our currently supported grant types does not need to be dynamic)

dteleguin committed:
rebase

dteleguin committed:
simplify OAuth2GrantType.Context creation

dteleguin committed:
Revert dynamic grant type resolution

dteleguin committed:
Split OAuth2GrantType and OAuth2GrantTypeFactory

dteleguin committed:
Fix UMA

dteleguin committed:
Fix disabled feature tests

tnorimat commented:
@dteleguin I am afraid that you missed [DCO](https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md#developers-certificate-of-origin), which causes a failure of CI processes.

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java:
```diff
@@ -154,9 +91,10 @@ private enum Action {
     private String grantType;
+    private List<OAuth2GrantType> grants;
+    private OAuth2GrantType grant;
+    private OAuth2GrantType.Context context;
```
This is not yet fully changed, but I don't consider this as a blocker of this PR (`context` can be likely declared as "method scoped"  variable instead of the "object scoped" variable. The `grant` can be a bit more tricky as it is currently used in more methods, so it is completely fine to keep it as it is IMO).

mposolda commented:
@dteleguin Thanks for the updates! Looks much better IMO.

There are still few comments pending like:

1) Remove method `setContext(Context)` from `OAuth2GrantType` and instead have method `process(Context)` . So `Context` is sent directly as an argument to `process` method

2) Possibly update the fields in class `OAuth2GrantType.Context` to be `final` 

3) Some other declared minor things inline, but I don't consider those as blockers for this PR.

Ideally, if (1) and (2) can be improved, we should be hopefully fine with merging this.

@tnorimat I don't consider DCO as blocker of this PR as I can squash commits in this PR before the merge, so the DCO requirement would be fulfilled .

dteleguin committed:
Remove setContext() + minor cleanup

Signed-off-by: Dmitry Telegin <demetrio@carretti.pro>

mposolda reviewed (approved):
@dteleguin Thanks! Maybe a little bit better option might be to add method like this to the OAuth2GrantBase:
```
public void process(Context context) {
    setContext(context);
    processImpl();
}

protected abstract processImpl();
```
and then override only `processImpl` in all the subclasses, so they don't need to always call `setContext(context)` at the beginning of method `process(context)` .

But I consider this as a minor point and approving this PR, so we catch this for Keycloak 24. We can possibly consider of polish later if needed? @pedroigor WDYT?

pedroigor commented on server-spi-private/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantType.java:
```diff
@@ -0,0 +1,139 @@
+            return session;
+        }
+
+        public void setFormParams(MultivaluedHashMap<String, String> formParams) {
```
We can have a follow-up on this one.

pedroigor reviewed (approved):
@dteleguin Nice!

dteleguin commented:
thanks everybody :partying_face:

dteleguin commented:
Maybe worth mentioning in the 24 release notes? It's not just about refactoring but also about supporting [extension grants](https://datatracker.ietf.org/doc/html/rfc6749#section-4.5), an important OAuth 2.0 feature.
